src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs

```diff
@@ -695,9 +695,22 @@ internal void ExecutePackageDownload(string name, PackageVersion package, string
                             Name = dep.name
                         };
                     })
-                    .ToList()
-                    .ForEach(x => DownloadAndInstall(x, downloadPath));
+                    .Select(x => Download(x)).ToArray();
+                //wait for all downloads.
+                await Task.WhenAll(downloadTasks);
+
+                //when above downloads complete, start installing packages in dependency order.
+                foreach(var dep in dependencyVersionHeaders)
+                {
+                    var matchingDownload = downloadTasks.Where(x => x.Result.handle.Id == dep.id).FirstOrDefault();
+                    if(matchingDownload != null)
```

Comment:
why do we need to match the download handles to the version headers ?
Is there a chance that a dependency did not download correctly ? If that is the case, will the error be handled as before ?

Reply:
@pinzart90 I'll add some comments but heres the deal:

1. we match the download handles to the version headers because the version headers are in topological order and the downloads are in any random order based on what completed first.
2. errors are handled inside the download as before - they return a handle in `error` state - this gets passed on to `InstallPackage` but it handles this and just returns - theres a test for this.